dashboard/modules/reporter/profile_manager.py

```diff
@@ -131,6 +144,125 @@ async def cpu_profile(
         )
         stdout, stderr = await process.communicate()
         if process.returncode != 0:
-            return False, _format_failed_pyspy_command(cmd, stdout, stderr)
+            return False, _format_failed_profiler_command(
+                cmd, self.profiler, stdout, stderr
+            )
         else:
             return True, open(profile_file_path, "rb").read()
+
+
+class MemoryProfilingManager:
+    def __init__(self, profile_dir_path: str):
+        self.profile_dir_path = Path(profile_dir_path) / "memray"
+        self.profile_dir_path.mkdir(exist_ok=True)
+        self.profiler = "memray"
+
+        if not self.profile_dir_path.exists():
+            self.profile_dir_path.mkdir()
+
+    async def get_profile_result(
+        self, pid: int, profiler_filename: str, format="flamegraph", leaks=False
+    ) -> (bool, str):
+        memray = shutil.which(self.profiler)
+        if memray is None:
+            return False, "memray is not installed"
+
+        profile_file_path = self.profile_dir_path / profiler_filename
+        if not Path(profile_file_path).is_file():
+            return False, f"process {pid} has not been profiled"
+
+        profiler_name, _ = os.path.splitext(profiler_filename)
+        profile_visualize_path = self.profile_dir_path / f"{profiler_name}.html"
+        if format == "flamegraph":
+            visualize_cmd = [
+                memray,
+                "flamegraph",
+                "-o",
+                profile_visualize_path,
+                "-f",
+            ]
+        elif format == "table":
+            visualize_cmd = [
+                memray,
+                "table",
+                "-o",
+                profile_visualize_path,
+                "-f",
+            ]
+        else:
+            return False, f"Report with format: {format} is not supported"
+
+        if leaks:
+            visualize_cmd.append("--leaks")
+        visualize_cmd.append(profile_file_path)
+
+        process = await asyncio.create_subprocess_exec(
+            *visualize_cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        stdout, stderr = await process.communicate()
+        if process.returncode != 0:
+            return False, _format_failed_profiler_command(
+                visualize_cmd, self.profiler, stdout, stderr
+            )
+
+        return True, open(profile_visualize_path, "rb").read()
+
+    async def attach_profiler(self, pid: int, native: bool = False) -> (bool, str):
```

Comment:
Can you handle the case where attach is requested again before it is finished? 

Also add tests for that case.